controller_manager:
  ros__parameters:
    update_rate: 125  # Hz

    joint_state_broadcaster:
      type: joint_state_broadcaster/JointStateBroadcaster

    io_and_status_controller:
      type: ur_controllers/GPIOController

    speed_scaling_state_broadcaster:
      type: ur_controllers/SpeedScalingStateBroadcaster

    force_torque_sensor_broadcaster:
      type: force_torque_sensor_broadcaster/ForceTorqueSensorBroadcaster

    joint_trajectory_controller:
      type: joint_trajectory_controller/JointTrajectoryController

    scaled_joint_trajectory_controller:
      type: ur_controllers/ScaledJointTrajectoryController

    forward_velocity_controller:
      type: velocity_controllers/JointGroupVelocityController

    forward_position_controller:
      type: position_controllers/JointGroupPositionController

    my_cartesian_motion_controller:
      type: "cartesian_motion_controller/CartesianMotionController"
    
    pid:
      type: "pid_controller/PidController"

    forward_velocity_controller_pid:
      type: forward_command_controller/ForwardCommandController


pid:
  ros__parameters:
    dof_names:
      - shoulder_pan_joint
      - shoulder_lift_joint
      - elbow_joint
      - wrist_1_joint
      - wrist_2_joint
      - wrist_3_joint

    command_interface: position

    reference_and_state_interfaces:
      # - position
      - velocity

    update_rate: 125

    gains:
      shoulder_pan_joint:  {"p": 1.0, "i": 1.0, "d": 0.0, "i_clamp_min": -7.0, "i_clamp_max": 7.0, "antiwindup": true}
      shoulder_lift_joint: {"p": 1.0, "i": 1.0, "d": 0.0, "i_clamp_min": -7.0, "i_clamp_max": 7.0, "antiwindup": true}
      elbow_joint:         {"p": 1.0, "i": 1.0, "d": 0.0, "i_clamp_min": -7.0, "i_clamp_max": 7.0, "antiwindup": true}
      wrist_1_joint:       {"p": 1.0, "i": 1.0, "d": 0.0, "i_clamp_min": -7.0, "i_clamp_max": 7.0, "antiwindup": true}
      wrist_2_joint:       {"p": 1.0, "i": 1.0, "d": 0.0, "i_clamp_min": -7.0, "i_clamp_max": 7.0, "antiwindup": true}
      wrist_3_joint:       {"p": 1.0, "i": 1.0, "d": 0.0, "i_clamp_min": -7.0, "i_clamp_max": 7.0, "antiwindup": true}

# preceeding_forward_velocity_controller_pid:
#   ros__parameters:
#     joints:
#       - pid/shoulder_pan_joint
#       - pid/shoulder_lift_joint
#       - pid/elbow_joint
#       - pid/wrist_1_joint
#       - pid/wrist_2_joint
#       - pid/wrist_3_joint
#     interface_name: velocity

forward_velocity_controller_pid:
  ros__parameters:
    joints:
      - pid/shoulder_pan_joint
      - pid/shoulder_lift_joint
      - pid/elbow_joint
      - pid/wrist_1_joint
      - pid/wrist_2_joint
      - pid/wrist_3_joint
    interface_name: velocity

speed_scaling_state_broadcaster:
  ros__parameters:
    state_publish_rate: 100.0

force_torque_sensor_broadcaster:
  ros__parameters:
    sensor_name: tcp_fts_sensor
    state_interface_names:
      - force.x
      - force.y
      - force.z
      - torque.x
      - torque.y
      - torque.z
    frame_id: tool0
    topic_name: ft_data

joint_trajectory_controller:
  ros__parameters:
    joints:
      - shoulder_pan_joint
      - shoulder_lift_joint
      - elbow_joint
      - wrist_1_joint
      - wrist_2_joint
      - wrist_3_joint
    command_interfaces:
      - position
    state_interfaces:
      - position
      - velocity
    state_publish_rate: 125.0
    action_monitor_rate: 20.0
    allow_partial_joints_goal: true
    open_loop_control: true
    constraints:
      stopped_velocity_tolerance: 0.1
      goal_time: 0.0
      shoulder_pan_joint: { trajectory: 10.1, goal: 0.1 }
      shoulder_lift_joint: { trajectory: 10.1, goal: 0.1 }
      elbow_joint: { trajectory: 10.1, goal: 0.1 }
      wrist_1_joint: { trajectory: 10.1, goal: 0.1 }
      wrist_2_joint: { trajectory: 10.1, goal: 0.1 }
      wrist_3_joint: { trajectory: 10.1, goal: 0.1 }
    gains:
      shoulder_pan_joint: {p: 10.0, i: 1.0, d: 0.1}
      shoulder_lift_joint: {p: 10.0, i: 1.0, d: 0.1}
      elbow_joint: {p: 10.0, i: 1.0, d: 0.1}
      wrist_1_joint: {p: 10.0, i: 1.0, d: 0.1}
      wrist_2_joint: {p: 10.0, i: 1.0, d: 0.1}
      wrist_3_joint: {p: 10.0, i: 1.0, d: 0.1}

forward_velocity_controller:
  ros__parameters:
    joints:
      - shoulder_pan_joint
      - shoulder_lift_joint
      - elbow_joint
      - wrist_1_joint
      - wrist_2_joint
      - wrist_3_joint
    # interface_name: velocity
    
    # update_rate: 125

forward_position_controller:
  ros__parameters:
    joints:
      - shoulder_pan_joint
      - shoulder_lift_joint
      - elbow_joint
      - wrist_1_joint
      - wrist_2_joint
      - wrist_3_joint

my_cartesian_motion_controller:
  ros__parameters:
    end_effector_link: "tool0"
    robot_base_link: "base_link"
    #NOTUSED target_frame
    target_frame_topic: "target_frame"
    joints:
    - shoulder_pan_joint
    - shoulder_lift_joint
    - elbow_joint
    - wrist_1_joint
    - wrist_2_joint
    - wrist_3_joint
    
    command_interfaces:
      - position

    pd_gains:
        trans_x: {p: 100.0}
        trans_y: {p: 100.0}
        trans_z: {p: 100.0}
        rot_x: {p: 20.0}
        rot_y: {p: 20.0}
        rot_z: {p: 20.0}
